### ISSUE TITLE
IDA Cloud Decompiler Not Working

### ISSUE THREAD

**Issue by ConnorNelson** [pwncollege/dojo#490]
`LD_LIBRARY_PATH=/nix/store/gkx5bfa7kf83wgh94s6vr93vn356w081-openssl-3.0.14/lib ida64` makes it work.

My attempt to fix did not work:
```
  ida-free = pkgs.ida-free.overrideAttrs (oldAttrs: {
    runtimeDependencies = oldAttrs.runtimeDependencies // [ (pkgs.lib.getLib pkgs.openssl) ];
  });
```

This seems to correctly bake into the `RUNPATH` (`readelf -a ... | grep RUNPATH` for both the opt/ida64 and opt/.ida64-wrapped), but for some reason this does not fix it.

`LD_DEBUG=libs ida64` does not show it trying to search the `RUNPATH`, but it does search `LD_LIBRARY_PATH`.

**Comment by ConnorNelson** [comment]
Related discussion in https://github.com/NixOS/nixpkgs/pull/299777.

**Comment by ConnorNelson** [comment]
https://stackoverflow.com/questions/53286707/why-would-runpath-be-ignored mentions that `dlopen` does not actually consider `RUNPATH`.

**Comment by Scoder12** [comment]
Adding a wrapper that sets LD_LIBRARY_PATH is an option.

**Comment by ConnorNelson** [comment]
Yeah, I went down the massive rabbit hole of understanding what's going on and it seems like that is truly our only option. Not totally sure if that is the "most correct" solution, but it is what I have settled on. It seems like the nix community has walked away from RPATH in favor of RUNPATH as a result of RPATH being deprecated (my brief understanding is that RPATH couldn't be overriden by `LD_LIBRARY_PATH`). It seems like there is plenty of precedent for nixpkgs being fixed up in this way.

**Comment by ConnorNelson** [comment]
@Scoder12 do you have any idea what the process is for upstreaming this fix to nixpkgs?

**Comment by Scoder12** [comment]
> It seems like there is plenty of precedent for nixpkgs being fixed up in this way.

Yeah, I've seen `LD_LIBRARY_PATH` wrappers before, especially with proprietary packages (discord desktop is one example). 

> do you have any idea what the process is for upstreaming this fix to nixpkgs?

Making a PR to nixpkgs. It should be single squashed commit and prefix the commit message with `ida-free: `
If you want to make yourself a maintainer of the package, you have to do a separate commit to add yourself to the maintainers list first. 
I can do this if you want since I already have a clone of nixpkgs hanging around -- should be a one line change?

**Comment by ConnorNelson** [comment]
Yeah, thanks!

The only thing I'm uncertain of is if `(lib.getLib openssl)` needs to be added as a dependency, or if openssl already provides the "library". I know that they do show up in separate stores, but presumably `openssl` brings it in? Or does just referencing `(lib.getLib openssl)` automatically bring it in? I don't really have a good mental model on this.

**Comment by Scoder12** [comment]
> does just referencing `(lib.getLib openssl)` automatically bring it in?

Referencing the store path in any way is enough for it to be a dependency of that package. If you want good mental models the "Nix Pills" series is good: [Pill 9 - Automatic Runtime Dependencies](https://nixos.org/guides/nix-pills/09-automatic-runtime-dependencies). 

> presumably `openssl` brings it in? 

Probably a safe bet that the binary package `openssl` (or `openssl.bin`) pulls in the library, but if you want access to transitive dependencies you generally have to reference them -- even though it exists in the store somewhere, you have to explicitly add it to `PATH` or `LD_LIBRARY_PATH` or whatever else.
